runtime/master/src/main/java/edu/snu/nemo/runtime/master/resource/ExecutorRepresenter.java

```diff
@@ -133,19 +136,41 @@ public void sendControlMessage(final ControlMessage.Message message) {
    *
    */
   public void onTaskExecutionComplete(final String taskId) {
-    runningTasks.remove(taskId);
-    runningTaskToAttempt.remove(taskId);
-    completeTasks.add(taskId);
+    Task completedTask = null;
+    for (Task task : runningTasks) {
+      if (task.getTaskId().equals(taskId)) {
+        completedTask = task;
+      }
+    }
+
+    if (completedTask == null) {
+      throw new RuntimeException("Completed task not found in its ExecutorRepresenter");
+    }
+
+    runningTasks.remove(completedTask);
```

Comment:
Replace L139-L148 with the following? (using the Java Collection interface, instead of manually looping through the collection)

if (!runningTasks.remove(completedTask)) {
  throw new RuntimeException("Completed task not found in its ExecutorRepresenter");
}